### ISSUE TITLE
Poor text viewer performance

### ISSUE THREAD

**Issue by VioletGiraffe** [VioletGiraffe/file-commander#59]
This ASCII text file takes WAY too long to open: https://drive.google.com/file/d/0B8Qc2atC6FVKUE9VSFZsYjhVSzA/view?usp=sharing

**Comment by VioletGiraffe** [comment]
`QTextBrowser::setPlainText` and `QTextBrowser::setSource(QUrl::fromLocalFile())` both take huge amounts of time and memory.

**Comment by antis81** [comment]
Try using either `QPlainTextEdit` (displays the example file in < 1s) or a `QWebView` instead. As an idea: Switch the view depending on the mime-type of the file.

**Comment by VioletGiraffe** [comment]
I have tried `QPlainTextEdit`, and it's even **worse**. Yes, it does load the text very quickly, but try closing the window. It takes *enormous* amount of time to free its memory, like 12 seconds for the example file.

**Comment by antis81** [comment]
I did try `QPlainTextEdit` and it's ultimately fast (like expected). Basically this is what you **always** want, if rendering plain text. Maybe I'll peek into my good old XQueryEvaluator project. It is capable of loading large text files (>1GB) into a QPlainTextEdit.

I'm currently playing with `QWebView`, which takes ~28 secs. on `textBrowser.setContent(data, QLatin1String("text/plain"));`. This is about the same duration as with `QTextBrowser` - which is a dirty hack and should never be used :).

**Comment by VioletGiraffe** [comment]
Again: `QPlainTextEdit` has a huge problem with memory deallocation, at least on Windows. It's unusable.

**Comment by antis81** [comment]
Actually I don't understand why you're having a problem with using QPlainTextEdit. I can assure you, I've used it on Windows (XP / 7) with files far bigger than 14MB and it performed fine.

Anyway, this revealed a bug in my project, which I just fixed. Would you mind to download and compile XQueryEvaluator to check performance? You'll need to rename the test file to `1.xml` and set the path as "XML source file" (the combobox). After that, click on the "eye" icon, which will load and render the file. Clicking again will destroy the view. With highlighting enabled, it takes ~7s on my machine (likely to be faster on yours). After disabling the highlighter in the code, loading and rendering the file boils down to ~2.5s. Last not least, disabling the line numbers, we get to the ~1s, which I also measured in file-commander.

I also tested `QWebView` in file-commander and found, it loads reasonably fast with it's default method, while providing full flexibility of handling rendering rich text and even full bloated HTML. However, the text search becomes incompatible and needs to be reimplemented (see: QWebView::findText()).

```c++
// changed textBrowser to a QWebView
ui->textBrowser->load(QUrl::fromLocalFile(_sourceFilePath));
```

Don't be jealous: This line replaces all of the text viewer's loading code.

And ... like with QPlainTextEdit, the loading time boils down to even <1s. Yet  it takes some 3-4 secs. until the document is shown. Anyway, this is acceptable isn't it?

**Comment by VioletGiraffe** [comment]
It sure is better than what we have right now, and I actually use the text viewer a lot so I need it to be fast. I'm not fond of introducing the `QWebView` dependency, though. But it's something to consider.
All my encoding detection magic will be gone, though. Does `QWebView` does the same? I'm not sure it will.
I'll try `XQueryEvaluator` soon.

**Comment by VioletGiraffe** [comment]
Actually, what's XQueryEvaluator? You never gave a link.

**Comment by antis81** [comment]
>Actually, what's XQueryEvaluator? You never gave a link.

Ah sorry. It's an editor (actually the only available open source) to ... well ... write and evaluate queries on XML files with the somewhat rare XQuery language - SQL for XML :smile:. Usually it expects the source to be an XML file, but you can feed it any text file. Source can be downloaded from [here](https://github.com/antis81/XQueryEvaluator/releases) and should compile fine with your setup. Every single text editor/viewer used is based on a QPlainTextEdit and I'm interested in how it performs on Windows.

**Comment by VioletGiraffe** [comment]
Fixed by #72.
